### ISSUE TITLE
Inline Parsing Clarification

### ISSUE THREAD

**Issue by cbreeden** [markdown-it/markdown-it#243]
I'm working on rewriting [markdown-it-katex](https://github.com/cbreeden/markdown-it-katex.git), partly as an exercise, but also with the goal to make the syntax compatible with [pandoc](http://pandoc.org/README.html#math)'s extension.  I also was curious in implementing an attribute system that followed the syntax discussed [here](http://talk.commonmark.org/t/consistent-attribute-syntax/272). I had a few questions to help clarify how inline parsing should work.

### Question

Concerning inline rulers, if my ruler is finished parsing at the current cursor position and decides that there should be no action, I figured that I should simply return false.  For instance, while looking for a closing $ delimiter, I figured I should write:

  ```js
  if (match == -1) { return false; }
  ```
  
  Unfortunately, I was only able to get my plugin to work with (cf [here](https://github.com/cbreeden/markdown-it-katex/blob/master/index.js#L54)):
  
  ```js
  if (match == -1) {
    if (!silent) { state.pending += "$"; }
    return true;
  }
  ```
  First of all, I'm not sure what the difference is.  After reading the [source](https://github.com/markdown-it/markdown-it/blob/master/lib/parser_inline.js#L145), i figured the difference was:  returning `false` will resume to check all other rulers, and if nothing happens then do a `state.pending += "$"`; doing a `state.pending += "$"` and then returning true tells markdown to bypass all other rulers, add `$` to pending (ie, tell markdown that this is text) and continue with inline parsing.  However some of my tests fail.

### Failed Test

This is an example failed test for when using `return false`.  When there is no `match`, the opening `$` not be displayed in the rendered contents. 

The test:

```
Shouldn't render USD
.
aaa $5.99 bbb
.
<p>aaa $5.99 bbb</p>
.
```

The result:

```
# Shouldn't render USD
not ok 6 should be equal
  ---
    operator: equal
    expected: '<p>aaa $5.99 bbb</p>\n'
    actual:   '<p>aaa 5.99 bbb</p>\n'
    at: Test.<anonymous> (/home/breeden/github/markdown-it-katex/test/all.js:22:6)
  ...
```

**Comment by puzrin** [comment]
Since math should not depend on other markup, you can fetch full range with formula at once. So, take a look at something like [backticks](https://github.com/markdown-it/markdown-it/blob/6.0.1/lib/rules_inline/backticks.js) or [entity](https://github.com/markdown-it/markdown-it/blob/6.0.1/lib/rules_inline/entity.js)

"silent" mode is used for detect next entires. It should change position on success (as usual) but should not generate output.

**Comment by cbreeden** [comment]
Reading through [backticks](https://github.com/markdown-it/markdown-it/blob/6.0.1/lib/rules_inline/backticks.js), I was able to conclude that it will only return `false` if there isn't a backtick character at the current cursor position.  So even if `backticks` determines that the backtick at the current cursor position isn't a valid opening sequence for code it will return true.  This is what I had, so I guess I can keep it that way.  

I guess I was just confused why it wasn't working when I return false, since I figured the inline parser essentially treated `return false;` and `if (!silent) { state.pending += "$"; } return true;` as the same, while I was looking at [this](https://github.com/markdown-it/markdown-it/blob/master/lib/parser_inline.js#L140):

```js
  while (state.pos < end) {
    if (state.level < maxNesting) {
      for (i = 0; i < len; i++) {
        ok = rules[i](state, false);
        if (ok) { break; }
      }
    }

    if (ok) {
      if (state.pos >= end) { break; }
      continue;
    }

    state.pending += state.src[state.pos++];
  }
```

**Comment by puzrin** [comment]
> So even if backticks determines that the backtick at the current cursor position isn't a valid opening sequence for code it will return true. This is what I had, so I guess I can keep it that way.

`true` means "have something correct". This rule can return "backtick found" or "plain text found" (if opening backtick is not matched).

**Comment by cbreeden** [comment]
So silly me.  I realized I modified the `state.pos` before returning false and the parser is expecting the cursor to be unchanged if I return false.  I believe this was my mistake and clarifies my understanding.  Sorry, I didn't expect this to be a trouble shooting ticket, I thought I genuinely had a misunderstanding of how I should handing inline-rulers. 

However, this statement helps me a lot, thank you!

> `true` means "have something correct". This rule can return "backtick found" or "plain text found" (if opening backtick is not matched).
